File: couple-carrier-risk/r/R/genome_file.R
# Validate a 23andMe raw genotype file and infer biological sex from it.
#
# The file is sent verbatim to Evagene's /import/23andme-raw endpoint;
# this module does *not* reinterpret genotypes or carrier state.  Its
# only jobs are to check the TSV is well-formed and to derive a
# biological_sex value we can record on the individual before import.

SEX_MALE <- "male"
SEX_FEMALE <- "female"
SEX_UNKNOWN <- "unknown"

EXPECTED_FIELDS <- 4L
Y_CHROMOSOME <- "Y"
NO_CALL <- "--"

#' Raise a genome_file_error condition.
#' @noRd
stop_genome_file <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("genome_file_error", "error", "condition")
  ))
}

#' Read, validate, and summarise a 23andMe raw genotype TSV.
#'
#' @param path Path to a 23andMe raw genotype file.
#' @return A named list with fields `path`, `content`, `biological_sex`.
load_genome_file <- function(path) {
  if (!file.exists(path)) {
    stop_genome_file(sprintf("23andMe file not found: %s", path))
  }
  content <- tryCatch(
    paste(readLines(path, warn = FALSE), collapse = "\n"),
    error = function(e) {
      stop_genome_file(sprintf("Cannot read 23andMe file %s: %s", path, conditionMessage(e)))
    }
  )

  rows <- parse_genotype_rows(content)
  if (length(rows$chromosome) == 0L) {
    stop_genome_file(sprintf(
      paste0("%s: no genotype rows found. ",
             "Expected a TSV with columns rsid, chromosome, position, genotype."),
      path
    ))
  }

  list(
    path = path,
    content = content,
    biological_sex = infer_sex(rows$chromosome, rows$genotype)
  )
}

parse_genotype_rows <- function(content) {
  chromosomes <- character(0)
  genotypes <- character(0)
  for (raw_line in strsplit(content, "\n", fixed = TRUE)[[1L]]) {
    line <- trimws(raw_line)
    if (!nzchar(line) || startsWith(line, "#")) next
    parts <- strsplit(line, "\t", fixed = TRUE)[[1L]]
    if (length(parts) < EXPECTED_FIELDS) next
    chromosomes <- c(chromosomes, parts[[2L]])
    genotypes <- c(genotypes, trimws(parts[[4L]]))
  }
  list(chromosome = chromosomes, genotype = genotypes)
}

#' Male if any Y-chromosome SNP has a real call; female if all are no-calls.
#'
#' 23andMe reports Y-chromosome SNPs as "--" for biological females (no
#' Y chromosome present).  When a file has no Y-chromosome rows at all
#' we cannot tell, and return "unknown".
infer_sex <- function(chromosomes, genotypes) {
  y_mask <- chromosomes == Y_CHROMOSOME
  if (!any(y_mask)) {
    return(SEX_UNKNOWN)
  }
  if (any(genotypes[y_mask] != NO_CALL)) {
    return(SEX_MALE)
  }
  SEX_FEMALE
}

File: couple-carrier-risk/python/src/couple_carrier_risk/genome_file.py
"""Validate a 23andMe raw genotype file and infer biological sex from it.

The file is sent verbatim to Evagene's ``/import/23andme-raw`` endpoint;
this module does *not* reinterpret genotypes or carrier state. Its only
jobs are to check the TSV is well-formed and to derive a ``biological_sex``
value we can record on the individual before import.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from pathlib import Path

_EXPECTED_FIELDS = 4
_Y_CHROMOSOME = "Y"
_NO_CALL = "--"


class BiologicalSex(str, Enum):
    MALE = "male"
    FEMALE = "female"
    UNKNOWN = "unknown"


class GenomeFileError(ValueError):
    """Raised when a 23andMe raw file is missing, empty, or malformed."""


@dataclass(frozen=True)
class GenomeFile:
    path: str
    content: str
    biological_sex: BiologicalSex


def load_genome_file(path: str | Path) -> GenomeFile:
    """Read, validate, and summarise a 23andMe raw genotype TSV.

    Raises :class:`GenomeFileError` if the file is missing, empty, or has
    no recognisable genotype rows. Non-fatal oddities (occasional short
    lines, comments, blank lines) are silently tolerated — 23andMe adds
    header notes that should not fail the import.
    """
    resolved = Path(path)
    try:
        content = resolved.read_text(encoding="utf-8")
    except FileNotFoundError as error:
        raise GenomeFileError(f"23andMe file not found: {resolved}") from error
    except OSError as error:
        raise GenomeFileError(f"Cannot read 23andMe file {resolved}: {error}") from error

    rows = list(_iter_genotype_rows(content))
    if not rows:
        raise GenomeFileError(
            f"{resolved}: no genotype rows found. Expected a TSV with columns "
            "rsid, chromosome, position, genotype.",
        )

    return GenomeFile(
        path=str(resolved),
        content=content,
        biological_sex=_infer_sex(rows),
    )


@dataclass(frozen=True)
class _Row:
    chromosome: str
    genotype: str


def _iter_genotype_rows(content: str) -> list[_Row]:
    rows: list[_Row] = []
    for raw_line in content.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        parts = line.split("\t")
        if len(parts) < _EXPECTED_FIELDS:
            continue
        rows.append(_Row(chromosome=parts[1], genotype=parts[3].strip()))
    return rows


def _infer_sex(rows: list[_Row]) -> BiologicalSex:
    """Male if any Y-chromosome SNP has a real call; female if all are no-calls.

    23andMe reports Y-chromosome SNPs as ``--`` for biological females
    (no Y chromosome present). When a file has no Y-chromosome rows at
    all we cannot tell, and return :data:`BiologicalSex.UNKNOWN`.
    """
    y_rows = [row for row in rows if row.chromosome == _Y_CHROMOSOME]
    if not y_rows:
        return BiologicalSex.UNKNOWN
    if any(row.genotype != _NO_CALL for row in y_rows):
        return BiologicalSex.MALE
    return BiologicalSex.FEMALE
